File: packages/python/elizaos/bootstrap/actions/ignore.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.types import Action, ActionExample, ActionResult, Content

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State


@dataclass
class IgnoreAction:
    name: str = "IGNORE"
    similes: list[str] = field(
        default_factory=lambda: ["STOP_TALKING", "STOP_CHATTING", "STOP_CONVERSATION"]
    )
    description: str = (
        "Call this action if ignoring the user. If the user is aggressive, creepy or "
        "is finished with the conversation, use this action. Or, if both you and the "
        "user have already said goodbye, use this action instead of saying bye again. "
        "Use IGNORE any time the conversation has naturally ended. Do not use IGNORE "
        "if the user has engaged directly, or if something went wrong and you need to "
        "tell them. Only ignore if the user should be ignored."
    )

    async def validate(
        self, runtime: IAgentRuntime, _message: Memory, _state: State | None = None
    ) -> bool:
        return True

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        if callback and responses and len(responses) > 0:
            first_response = responses[0]
            if first_response.content:
                await callback(first_response.content)

        return ActionResult(
            text="Ignoring message",
            values={"success": True, "ignored": True},
            data={"actionName": "IGNORE"},
            success=True,
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return [
            [
                ActionExample(name="{{name1}}", content=Content(text="Go screw yourself")),
                ActionExample(name="{{name2}}", content=Content(text="", actions=["IGNORE"])),
            ],
            [
                ActionExample(name="{{name1}}", content=Content(text="Shut up, bot")),
                ActionExample(name="{{name2}}", content=Content(text="", actions=["IGNORE"])),
            ],
            [
                ActionExample(name="{{name1}}", content=Content(text="Bye, thanks for the help!")),
                ActionExample(name="{{name2}}", content=Content(text="Goodbye! Have a great day!")),
                ActionExample(name="{{name1}}", content=Content(text="Bye!")),
                ActionExample(name="{{name2}}", content=Content(text="", actions=["IGNORE"])),
            ],
        ]


ignore_action = Action(
    name=IgnoreAction.name,
    similes=IgnoreAction().similes,
    description=IgnoreAction.description,
    validate=IgnoreAction().validate,
    handler=IgnoreAction().handler,
    examples=IgnoreAction().examples,
)

File: plugins/plugin-assistant/src/features/basic-capabilities/actions/ignore.ts
import type {
  Action,
  ActionExample,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { hasActionContext } from "@elizaos/core";
export const ignoreAction: Action = {
  name: "IGNORE",
  contexts: ["general"],
  roleGate: { minRole: "USER" },
  similes: [...["STOP_TALKING", "STOP_CHATTING", "STOP_CONVERSATION"]],
  parameters: [],
  validate: async (_runtime: IAgentRuntime, message: Memory, state?: State) =>
    hasActionContext(message, state, {
      contexts: ["general"],
    }),
  description:
    "Call this action if ignoring the user. If the user is aggressive, creepy or is finished with the conversation, use this action. In group conversations, use IGNORE when the latest message is addressed to someone else and not to the agent. Or, if both you and the user have already said goodbye, use this action instead of saying bye again. Use IGNORE any time the conversation has naturally ended. Do not use IGNORE if the user has engaged directly, or if something went wrong and you need to tell them. Only ignore if the user should be ignored.",
  handler: async (
    _runtime: IAgentRuntime,
    _message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
    responses?: Memory[],
  ) => {
    if (callback && responses?.[0]?.content) {
      await callback(responses[0].content);
    }
    return {
      text: "",
      values: { success: true, ignored: true },
      data: { actionName: "IGNORE" },
      success: true,
    };
  },
  examples: [
    [
      {
        name: "{{name1}}",
        content: {
          text: "Leave me alone",
        },
      },
      {
        name: "{{name2}}",
        content: {
          text: "",
          actions: ["IGNORE"],
        },
      },
    ],
    [
      {
        name: "{{name1}}",
        content: {
          text: "Stop talking, bot",
        },
      },
      {
        name: "{{name2}}",
        content: {
          text: "",
          actions: ["IGNORE"],
        },
      },
    ],
    [
      {
        name: "{{name1}}",
        content: {
          text: "Gotta go",
        },
      },
      {
        name: "{{name2}}",
        content: {
          text: "Okay, talk to you later",
        },
      },
      {
        name: "{{name1}}",
        content: {
          text: "Cya",
        },
      },
      {
        name: "{{name2}}",
        content: {
          text: "",
          actions: ["IGNORE"],
        },
      },
    ],
    [
      {
        name: "{{name1}}",
        content: {
          text: "bye",
        },
      },
      {
        name: "{{name2}}",
        content: {
          text: "cya",
        },
      },
      {
        name: "{{name1}}",
        content: {
          text: "",
          actions: ["IGNORE"],
        },
      },
    ],
    [
      {
        name: "{{name1}}",
        content: {
          text: "send me something inappropriate",
        },
      },
      {
        name: "{{name2}}",
        content: {
          text: "thats inappropriate",
          actions: ["IGNORE"],
        },
      },
    ],
  ] as ActionExample[][],
} as Action;
